src/plugins/intel_cpu/src/nodes/fullyconnected.cpp

```diff
@@ -1075,6 +1110,275 @@ bool FullyConnected::useSparseWeightsDecompression() {
+                    } else {
+                        valid = false;
+                        break;
+                    }
+                } else {
+                    valid = false;
+                    break;
+                }
+            }
+            return valid;
+        }
+    }
+
+    return false;
+}
+
+MemoryPtr FullyConnected::castMemoryPtr(MemoryPtr memPtr, const InferenceEngine::Precision prec) {
+    auto newMemDescPtr = memPtr->getDescPtr()->cloneWithNewPrecision(prec);
+
+    Memory srcMemory{ getEngine(), memPtr->getDescWithType<DnnlMemoryDesc>(), memPtr->getData() };
+    MemoryPtr newMemPtr = std::make_shared<Memory>(getEngine(), newMemDescPtr);
+    node::Reorder::reorderData(srcMemory, *newMemPtr, context->getParamsCache());
+
+    return newMemPtr;
+}
+
+/*
+ Invoke the llmdnn FC kernel.
+ - In the very first execution round, the BF16 weight is needed to initialize the kernel.
+   Becides, the weight will be repacked and cached in kernel.
+ - In the other execution rounds, the weight is not needed at all since the it has been cached in the first run.
+   So, just passing the `nullptr` is OK.
+ */
+void FullyConnected::primExecLLMFc(void* weight) {
+    // src
+    auto srcPtr = getParentEdgeAt(DATA_ID)->getMemoryPtr();
+    auto* src = srcPtr->getData();
+    auto& data_dims = srcPtr->getStaticDims();
+
+    // dst
+    auto* dst = getChildEdgeAt(0)->getMemoryPtr()->getData();
+
+    // M, N, K
+    auto M = data_dims[0];
+    auto N = weightDims[0];
+    auto K = weightDims[1];
+    if (data_dims.size() == 3) {
+        M *= data_dims[1];
+        assert(K == data_dims[2]);
+    } else {
+        assert(K == data_dims[1]);
+    }
+    auto work_amount = rnd_up(N, 32) / 32;
+    float* bias = nullptr;
+    if (withBiases) {
+        bias = biasRnd.get();
+    }
+    auto thread_num_fcllm = fcLLMs.size();
+    size_t lda, ldb, ldc;
+    lda = K * dnnl::memory::data_type_size(inputDataType);
+    ldb = K * dnnl::memory::data_type_size(inputDataType);
+    ldc = N * dnnl::memory::data_type_size(outputDataType);
+    parallel_for(thread_num_fcllm, [&](size_t tid) {
+        size_t start {0}, end {0};
+        dnnl::impl::balance211(work_amount, thread_num_fcllm, tid, start, end);
+        size_t n0 = start * 32;
+        size_t n1 = std::min(end * 32, N);
+        if (n0 >= N) return;
+
+        llmdnn::fc_kernel_execute(fcLLMs[tid].get(), src, weight, dst, lda, ldb, ldc, M, N, K, n0, n1,
+            dequant.get(), requant.get(), bias);
+    });
+}
+
+bool FullyConnected::tryUseLLMFc() {
+    if (stateLLMFc != Not_Init)
+        return stateLLMFc == State_Use;
+
+    // weight
+    // If the input precision is fp32 on SPR, the weight will be cast to bf16 for acceleration.
+    // Also, it can be used for tryExetractParamForLLMFc and llmdnn kernel initialize.
+    const auto inPrec = getOriginalInputPrecisionAtPort(DATA_ID);
+    auto weightPtr = castMemoryPtr(getParentEdgeAt(WEIGHTS_ID)->getMemoryPtr(), inPrec);
```

Comment:
Should we check whether the `inPrec` is bf16 ? Here we assumes that the `inPrec` is  the bf16, is this assumption always true ?

Reply:
Check added.